### ISSUE TITLE
502 Bad Gateway errors generated by gorouter are not logged under responses_5xx

### ISSUE THREAD

**Issue by stephendotcarter** [cloudfoundry/gorouter#158]
`502 Bad Gateway` errors generated by gorouter are not logged under `responses_5xx`.

This leads to the `/varz` metrics not adding up correctly

Example:

```
requests:      510
responses_2xx: 312
responses_3xx: 0
responses_4xx: 51
responses_5xx: 0
responses_xxx: 4
```

Total responses is `367` which means there were `143` (510-367) requests that did not have a response.

Calculating the `requests` VS `responses` is a good method to determine how many "pending" requests the gorouter is working on.
But due to this issue the calculation can lead to constant non-zero values.

This can be reproduced by stopping the cloud controller and trying to make requests to API endpoints. The gorouter will reply with HTTP 502s which will not be accounted for under `responses_5xx`

The gorouter responds to the client with a `502 Bad Gateway`

https://github.com/cloudfoundry/gorouter/blob/156e02996e28437e85bce7186ed8a4c2e074937c/proxy/proxy.go#L279

And then the code returns.

Even though this generates a 502 response to the client, `responses_5xx` does not incremented since the counter is not incremented until further down the code:
https://github.com/cloudfoundry/gorouter/blob/156e02996e28437e85bce7186ed8a4c2e074937c/proxy/proxy.go#L294

```
        if endpoint == nil {
            handler.HandleBadGateway(err, request) <-- return here due to 502
            return
        }

        accessLog.FirstByteAt = time.Now()
        if rsp != nil {
            accessLog.StatusCode = rsp.StatusCode
        }

        if p.traceKey != "" && endpoint != nil && request.Header.Get(router_http.VcapTraceHeader) == p.traceKey {
            router_http.SetTraceHeaders(responseWriter, p.ip, endpoint.CanonicalAddr())
        }

        latency := time.Since(accessLog.StartedAt)

        p.reporter.CaptureRoutingResponse(endpoint, rsp, accessLog.StartedAt, latency) <-- counter does not get incremented since this line is not reached
```

This leads to a situation where the metrics for `responses` do not sum up to the metrics for `requests` even though we sent a valid response.

This issue could also affect some other response codes that come directly from gorouter (i.e. Not an endpoint response)

**Comment by abbyachau** [comment]
Hello @stephendotcarter 
Thank you for looking into this. As `/varz` is deprecated, can you verify the same behavior can be observed using metrics available from the firehose?

http://docs.cloudfoundry.org/loggregator/all_metrics.html#gorouter

Thank you!

**Comment by stephendotcarter** [comment]
This issue was originally noticed from viewing the metrics through the JMX tile. But in troubleshooting and narrowing down the issue we noticed the same metrics mismatch in `/varz`.

Kind regards,
Stephen

**Comment by shashwathi** [comment]
@stephendotcarter: 

Thank you for raising this issue. You are correct `gorouter` does not capture 502 errors it generates under the metric`responses_5xx`. 
Reasoning behind this behavior is to help operator differentiate `Bad Gateway` errors generated from gorouter and backend instance. Metric called `bad_gateways` is used to capture `502` errors generated by `gorouter`. 

Regards
Shash

**Comment by stephendotcarter** [comment]
Thanks @shashwathi, what you are saying makes sense but I am still seeing a discrepancy.

This is output from a test environment.
```
requests         487232

bad_gateways         85
bad_requests         44
    
responses_2xx    385346
responses_3xx       914
responses_4xx    100630
responses_5xx        64
responses_xxx        85
```

Doing the math:
```
Total responses  487039
+ bad_gateways   487124
+ bad_requests   487168

Requests - everything else:
487232 - 487168 = 64
```

So there are still 64 requests unaccounted for.
Is my assumption correct that the numbers should add up?


Kind regards,
Stephen

**Comment by shashwathi** [comment]
Hi @stephendotcarter 

You are right. All the incoming requests should be accounted for in the metrics. @shalako @abbyachau have prioritized the story for engineering team to dig in more.

Regards
Shash

**Comment by abbyachau** [comment]
We've validated that release 0.146.0, included in cf-release v253 includes two new metrics for `origin: router`: `websocket_upgrades` and `websocket_failures` that account for the missing requests you identified. 

Please give it a try and get back to us. Thank you!

**Comment by abbyachau** [comment]
Hi @stephendotcarter, please see our note above - we've updated our metrics to include `origin: router`: `websocket_upgrades` and `websocket_failures`, which should account for the missing requests you've identified. This update was included in routing release 0.146.0. Please let us know if you have any feedback, thanks!

**Comment by abbyachau** [comment]
Hi @stephendotcarter, we are going to close this issue now due to inactivity but please do not hesitate to reopen or start a new issue if you have any feedback regarding the work we've done. Thanks again!
